From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ROCKNIX <rocknix@rocknix.org>
Date: Tue, 04 Mar 2026 00:00:00 +0000
Subject: [PATCH] arm64: dts: rockchip: rk356x: add DMC devfreq node and OPP
 table

Add the DMC (Dynamic Memory Controller) device tree node and DDR OPP
table to rk356x-base.dtsi.  The node is disabled by default; boards
must provide center-supply and set status = "okay" to enable DDR
frequency scaling.

The DMC driver (rk3568_dmc) uses the BSP V2 SIP protocol with MCU-based
DCF.  It reads DDR utilization from the DFI event device and adjusts
frequency via ATF shared-memory commands.

OPP table includes all known ATF FSP (Frequency Set Point) entries:
324/528/780/920/1056 MHz.  920 MHz is present on some RK3568 boards
(e.g. RGDS with DDR3).  ATF firmware will disable any OPP that does
not match a trained FSP at probe time, so unused entries are safe.

Signed-off-by: ROCKNIX <rocknix@rocknix.org>
---
 arch/arm64/boot/dts/rockchip/rk356x-base.dtsi | 40 +++++++++++++++++++++++
 1 file changed, 40 insertions(+)

diff --git a/arch/arm64/boot/dts/rockchip/rk356x-base.dtsi b/arch/arm64/boot/dts/rockchip/rk356x-base.dtsi
index XXXXXXX..XXXXXXX 100644
--- a/arch/arm64/boot/dts/rockchip/rk356x-base.dtsi
+++ b/arch/arm64/boot/dts/rockchip/rk356x-base.dtsi
@@ -1016,6 +1016,46 @@ dfi: dfi@fe230000 {
 		rockchip,pmu = <&pmugrf>;
 	};
 
+	dmc: memory-controller {
+		compatible = "rockchip,rk3568-dmc";
+		devfreq-events = <&dfi>;
+		clocks = <&cru CLK_DDR1X>;
+		clock-names = "dmc_clk";
+		operating-points-v2 = <&dmc_opp_table>;
+		interrupts = <GIC_SPI 10 IRQ_TYPE_LEVEL_HIGH>;
+		interrupt-names = "complete";
+		status = "disabled";
+	};
+
+	dmc_opp_table: opp-table-dmc {
+		compatible = "operating-points-v2";
+
+		opp-324000000 {
+			opp-hz = /bits/ 64 <324000000>;
+			opp-microvolt = <900000>;
+		};
+
+		opp-528000000 {
+			opp-hz = /bits/ 64 <528000000>;
+			opp-microvolt = <900000>;
+		};
+
+		opp-780000000 {
+			opp-hz = /bits/ 64 <780000000>;
+			opp-microvolt = <900000>;
+		};
+
+		opp-920000000 {
+			opp-hz = /bits/ 64 <920000000>;
+			opp-microvolt = <900000>;
+		};
+
+		opp-1056000000 {
+			opp-hz = /bits/ 64 <1056000000>;
+			opp-microvolt = <900000>;
+		};
+	};
+
 	pcie2x1: pcie@fe260000 {
 		compatible = "rockchip,rk3568-pcie";
 		reg = <0x3 0xc0000000 0x0 0x00400000>,
-- 
2.34.1

